### PR TITLE
tls_test: use a dedicated https server for testing

### PR DESCRIPTION
**Pull request by tchaikov**
the goal of this change set is to drop the dependency on an external server for performing unit tests. ideally, the unit tests are supposed to be self-contained. the behavior of external resources is unpredictable.

- drop a test relying on CA certs provided by system.
- instead of using google.com as an HTTPS server for testing, use a dedicated HTTPS server implemented using python

### CODE REVIEW

**Comment by tchaikov** [comment]
this change is a super set of #1346.

**Comment by tchaikov** [comment]
@elcallio hi Calle, could you help review this change?

File: tests/unit/tls_test.cc [hunk 1]
```diff
@@ -157,44 +186,54 @@ SEASTAR_TEST_CASE(test_x509_client_with_priority_strings) {
         "SECURE256:+SECURE128",
         "NORMAL:%COMPAT",
         "NORMAL:-MD5",
-        "NONE:+VERS-TLS-ALL:+MAC-ALL:+RSA:+AES-128-CBC:+SIGN-ALL:+COMP-NULL",
+        "NONE:+VERS-TLS-ALL:+MAC-ALL:+RSA:+AES-256-GCM:+SIGN-ALL:+COMP-NULL:+GROUP-ALL",
```
Comment (elcallio) [inline tests/unit/tls_test.cc:185]: Seems unrelated to rest of change.

File: tests/unit/tls_test.cc [hunk 2]
```diff
@@ -65,8 +65,8 @@ static std::string certfile(const std::string& file) {
 
 using namespace seastar;
 
-template <typename Func>
-auto with_https_server(Func&& func) {
+template <typename T, typename Func>
+auto with_https_server_and(T&& v, Func&& func) {
```
Comment (elcallio) [inline tests/unit/tls_test.cc:69]: Why not rebase this change into the first one?

File: tests/unit/tls_test.cc [hunk 3]
```diff
@@ -158,20 +158,20 @@ SEASTAR_TEST_CASE(test_simple_x509_client) {
 
 SEASTAR_TEST_CASE(test_x509_client_with_builder) {
     return with_https_server_and(tls::credentials_builder{}, [](socket_address addr, std::string cert, auto& b) {
-            return b.set_x509_trust_file(cert, tls::x509_crt_format::PEM).then([&b, addr] {
-                return connect_to_ssl_addr(b.build_certificate_credentials(), addr);
-            });
+        return b.set_x509_trust_file(cert, tls::x509_crt_format::PEM).then([&b, addr] {
+            return connect_to_ssl_addr(b.build_certificate_credentials(), addr);
+        });
```
Comment (elcallio) [inline tests/unit/tls_test.cc:163]: Should be done, but seems unrelated to the overall patch set.

**Comment by elcallio** [comment]
So I get where you are coming from, and on one hand I agree; Tests should be self-contained if possible. On the other hand there is a method to the insanity in the calls to google + using system trust; Simply testing that interoping with "real" world runs as expected. Since these will break in some build cases (isolated) I would suggest you instead of removing these tests do the following:

1.) Make you changes _additional_ tests instead (basically duplicates but using local server). This has an additional benefit of having a second test vector; local vs. "real" network stuff (with all the delays the latter add). 
2.) Add a preprocessor +token  cmake switch to allow disabling the "non-local" tests from build if needed. That way a seastar builder can opt out from tests that require wire.

One of the reasons I would like to keep the google stuff etc is that these tests (due to network delays etc) literally _did_ find edge cases in some things.

**Review by elcallio — CHANGES_REQUESTED** [review]

**Comment by avikivity** [comment]
@elcallio note that google is not available in all parts of the world.

**Comment by elcallio** [comment]
Which is why I suggest allowing it to be disabled at config time. Can even add a preproc-define + cmake switch to pick the public "auth" host, defaulting to google (or icann for that matter).

**Comment by avikivity** [comment]
> Which is why I suggest allowing it to be disabled at config time. Can even add a preproc-define + cmake switch to pick the public "auth" host, defaulting to google (or icann for that matter).

How about a demo program in demos/ instead? The test can be self-contained and test edge cases (like bad certificates). The demo can show idiomatic use and accept a command-line parameter for the server to talk to.

File: tests/unit/tls_test.cc (hunk 2 above)
Comment (tchaikov) [reply tests/unit/tls_test.cc:69]: i wanted to reduce the pain of reviewing because of reindent only changes. will fold this change into the "[tls_test: use dedicated https server for testing](https://github.com/scylladb/seastar/pull/1348/commits/9b339d9585abdf23c5505b45100af9bf03c4caa7)" commit.

File: tests/unit/tls_test.cc (hunk 1 above)
Comment (tchaikov) [reply tests/unit/tls_test.cc:185]: it's related. i am quoting part of the commit message here

>     because the HTTPS server started by https-server.py uses ECDHE
>     for key exchange, we need to enable a elliptic curves group. please
>     note, according to
>     https://gnutls.org/manual/html_node/Priority-Strings.html,
>     NORMAL enables both ECDHE, DHE and plain RSA. and since ECDHE
>     is more secure than RSA, let's use ECDHE for testing.

Comment (elcallio) [reply tests/unit/tls_test.cc:185]: Fair enough, but adding "GROUP-ALL" just enabled all default things, and more or less makes the prio part before it pointless, no? Not really testing any specific combo at all...

**Comment by elcallio** [comment]
My main concern is that "self-contained" will again just mean we test only localhost+self-signed certs. Not TLS with adding actual network latencies, packet frag etc (that really had issues), nor actual certificate auth signed chains. Even such a thing as verifying that system trust works is kind of important for potential seastar users (using it w.o. self-managed certs/auth).

Again, we actually had bugs materialize using this and its corresponding DNS/bind counterpart. 

I 100% agree with the OP:s concerns, but I would still like to keep the WAN testing when possible. If you disagree, I have no other real gripes with this (apart from the nits above).

File: tests/unit/tls_test.cc (hunk 2 above)
Comment (elcallio) [reply tests/unit/tls_test.cc:69]: fair enough.

File: tests/unit/tls_test.cc (hunk 1 above)
Comment (tchaikov) [reply tests/unit/tls_test.cc:185]: > Fair enough, but adding "GROUP-ALL" just enabled all default things,

i don't think so. "GROUP-ALL" is not "NORMAL". it only enables all groups for ec and finite field groups.

File: tests/unit/tls_test.cc (hunk 3 above)
Comment (tchaikov) [reply tests/unit/tls_test.cc:163]: it's just how the certs are built. if we stick to system CA certs, this change would not be necessary. otherwise, we need to use `tls::credentials_builder`.

File: tests/unit/tls_test.cc (hunk 1 above)
Comment (elcallio) [reply tests/unit/tls_test.cc:185]: Docs: `Catch all which enables all groups from NORMAL priority is GROUP-ALL`. I may be wrong, but it sounds awfully like "NORMAL" to me...
`GROUP-EC-ALL` enables all EC variants.

File: tests/unit/tls_test.cc (hunk 3 above)
Comment (elcallio) [reply tests/unit/tls_test.cc:163]: Sorry, I thought this was re-indenting old code.

**Comment by avikivity** [comment]
We could inject latency (and packet boundaries) using loopback_socket.

**Comment by elcallio** [comment]
We could, and should. Would be a good extension to the tests added here (variants with varying breaking/fragment/latency, yet controlled). But it still does not replace things like "google servers return empty response iff caller delays to much", which while not exposing a tls bug as such, did raise some other general issues that were fun to realize. Just things in the wild and running on different (a.k.a. bennys) machine(s).

File: tests/unit/tls_test.cc (hunk 1 above)
Comment (tchaikov) [reply tests/unit/tls_test.cc:185]: > Docs: `Catch all which enables all groups from NORMAL priority is GROUP-ALL`. I may be wrong, but it sounds awfully like "NORMAL" to me...

please note, GROUP-ALL does not imply `KX-ALL` or `CIPHER-ALL`. again, it only enables the listed groups in the document.

**Comment by tchaikov** [comment]
@elcallio hi Calle, could you help re-review this change? i

- added a cmake option named `Seastar_TESTING_WITH_NETWORKING` to enable the tests requiring network access. we can apply it to DNS tests in a follow-up change.
- squash the commits into a single one
- preserve the tests accessing google.com, and guard them with a macro named `SEASTAR_TESTING_WITH_NETWORKING`
- the tests using local web servers are mutual exclusive with those accessing google.com

it turns out to unify these two set of tests make them more complicated than necessary, so i think we can just replicate them.

File: tests/unit/CMakeLists.txt [hunk 4]
```diff
@@ -634,8 +635,20 @@ add_custom_target(tls_files
   DEPENDS ${out_tls_certificate_files}
 )
 
+if (NOT Seastar_TESTING_WITH_NETWORKING)
```
Comment (elcallio) [inline tests/unit/CMakeLists.txt:638]: Would it not make sense to always include the !TESTING_WITH_NETWORKING and just filter out outcall tests?
I can't imagine they add much to run time, and more testing by default is good.

File: CMakeLists.txt [hunk 5]
```diff
@@ -275,6 +275,10 @@ option (Seastar_TESTING
   "Enable testing targets."
   ON)
 
+include (CMakeDependentOption)
+cmake_dependent_option (Seastar_TESTING_WITH_NETWORKING "Test with networking access" ON
```
Comment (elcallio) [inline CMakeLists.txt:279]: Is the idea to have this as a "general" option, i.e. try to remove any test calling external (wan) networks if this is OFF?
Otherwise it should maybe be named after the test group (tls)? I am ok with either intent, but I would prefer a more explanatory description like "Enable tests using network WAN access".

File: tests/unit/tls_test.cc [hunk 6]
```diff
@@ -104,6 +104,8 @@ static future<> connect_to_ssl_addr(::shared_ptr<tls::certificate_credentials> c
     }).discard_result();
 }
 
+#ifdef SEASTAR_TESTING_WITH_NETWORKING
```
Comment (elcallio) [inline tests/unit/tls_test.cc:108]: Should be #if SEASTAR_TESTING_WITH_NETWORKING so it can be set to 0 or 1

**Comment by elcallio** [comment]
A few nits, but looking good.

File: tests/unit/CMakeLists.txt (hunk 4 above)
Comment (tchaikov) [reply tests/unit/CMakeLists.txt:638]: i don't feel strong either way. just wanted to have a decent coverage with relatively small efforts. will always enable the local tests and disable the tests using network access with the new option.

File: CMakeLists.txt (hunk 5 above)
Comment (tchaikov) [reply CMakeLists.txt:279]: > Is the idea to have this as a "general" option, i.e. try to remove any test calling external (wan) networks if this is OFF?

yes. as i put in the cover letter of the 2nd revision, i plan to apply this option to DNS tests in a follow-up change.

> Otherwise it should maybe be named after the test group (tls)? I am ok with either intent, but I would prefer a more explanatory description like "Enable tests using network WAN access".

i'd prefer using a more widely used name.  see the manpage of mock

> [--enable-network](https://www.mankier.com/1/mock#--enable-network)
> Enable networking. If you want to have reproducible builds then your builds should run without a network. This option overrides config_opts['rpmbuild_networking'] and config_opts['use_host_resolv'], setting both True.

see https://www.mankier.com/1/mock

and the manpage of pbuilder

> [--use-network [yes|no]](https://manpages.debian.org/testing/pbuilder/pbuilder.8.en.html#use)
> Specify yes when you do not want to disable network access during build. Network is not available on a Debian buildd, so you might want to keep the default of no. Disabling network access currently only works on Linux.

see https://manpages.debian.org/testing/pbuilder/pbuilder.8.en.html

but if neither of these example is convincing enough, how about using "internet"? as i think, "WAN" is used in the context of describing a network of network (LAN), but what we want to describe is the general access to the external network.

Comment (elcallio) [reply CMakeLists.txt:279]: Yes, "enable tests accessing internet" seems like a very descriptive naming/description base. Maybe ENABLE_INTERNET_ACCESS_TESTS?
But regarding WAN vs. interwebs; The TLS tests obviously accesses internet proper, but the dns/bind tests don't explicitly do this, they just use the configured DNS to resolve internet addresses, which is somewhat different.

File: tests/unit/tls_test.cc (hunk 6 above)
Comment (tchaikov) [reply tests/unit/tls_test.cc:108]: sorry, i don't quite follow here. why is `#if` better than `#ifdef` ? also, i am trying to follow the convention of other macros like `SEASTAR_DEBUG`, `SEASTAR_COROUTINES_ENABLED` and `SEASTAR_HAVE_DPDK`. i think this macro will not be a public interface used by developers directly, they should use `Seastar_TESTING_WITH_NETWORKING`.

Comment (elcallio) [reply tests/unit/tls_test.cc:108]: Personal preference, but (imho) `#if <boolean define>` is better that `#ifdef <symbol>` simply because it allows easier build script management (a -D<define>=<value> only has to change value, not remove/add def). It is also more concise (ask value, not existence)
Here we are using cmake, so maybe no a huge issue, but even cmake ensures the define is set to 1 for the proproc -> define configure things.

Comment (tchaikov) [reply tests/unit/tls_test.cc]: okay. but we don't generate a header file using `configure_file()` yet.  anyway, i changed to `#if`. but i still think it'd be desirable to be more consistent with other macros.

**Comment by tchaikov** [comment]
@elcallio hi Calle, thank you for your suggestions! could you help review again?

- change the doc string of the new option to `Enable tests accessing internet`
- change the option name to `Seastar_ENABLE_TESTS_ACCESSING_INTERNET`
- always enable the tests accessing local HTTPS server 
- use `#if` instead of `#ifdef` for guarding the disabled tests

**Comment by elcallio** [comment]
Looks good to me. @avikivity?

File: tests/unit/https-server.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python3
```
Comment (avikivity) [inline tests/unit/https-server.py:1]: Copyright/license blurb missing.

File: tests/unit/tls_test.cc [hunk 8]
```diff
@@ -201,6 +202,125 @@ SEASTAR_TEST_CASE(test_x509_client_with_priority_strings_fail) {
 }
 #endif // SEASTAR_TESTING_WITH_NETWORKING
 
+template <typename T, typename Func>
+auto with_https_server_and(T&& v, Func&& func) {
```
Comment (avikivity) [inline tests/unit/tls_test.cc:206]: Can we detemplate it?

 - make the caller capture `v` itself (it can use SEASTAR_THREAD_TEST_CASE if it wants a stable address)
 - change Func to std::function<future<> (socket_address sa, std::string cert)>
 - change the return type to future<>

If not, then the parameters must be constrained as a way to tell the reader what to expect.

File: tests/unit/https-server.py (hunk 7 above)
Comment (tchaikov) [reply tests/unit/https-server.py:1]: added.

File: tests/unit/tls_test.cc (hunk 8 above)
Comment (tchaikov) [reply tests/unit/tls_test.cc]: sure. now that `SEASTAR_THREAD_TEST_CASE` is used for adding the new tests, i take the opportunity to refactor `with_https_server_and()` to a class for better readability. it looks more closer to a coroutinized version.

**Comment by tchaikov** [comment]
@elcallio, @avikivity  hi Avi and Calle, thanks for your reviews.

- add license and copyright header to the newly added `https-server.py`
- refactor `with_https_server_and()` to a class `https_server` to address the comment requesting to detemplate the template function, and to improve the readability
- replace the magic string of `"127.0.0.1"` with a getter of `https_server`, accompany it with a comment explaining its connection to the related command in the CMake script
- extract the basename of cert and key files to a parameter of the constructor of `https_server`, less repeatings this way.
- add a comment to document the reason why we are reading from the output of the script
